File: packages/plugin-sdk/assembly/websocket.ts
import {
  peekoo_websocket_close,
  peekoo_websocket_connect,
  peekoo_websocket_recv,
  peekoo_websocket_send,
} from "./host";
import { extractStringField, quote } from "./json";
import { readString, writeString } from "./memory";

export function connect(url: string): string {
  const input = "{\"url\":" + quote(url) + "}";
  const offset = peekoo_websocket_connect(writeString(input));
  return extractStringField(readString(offset), "socketId");
}

export function send(socketId: string, text: string): void {
  const input =
    "{\"socketId\":" + quote(socketId) + ",\"text\":" + quote(text) + "}";
  peekoo_websocket_send(writeString(input));
}

export function recv(socketId: string): string {
  const input = "{\"socketId\":" + quote(socketId) + "}";
  const offset = peekoo_websocket_recv(writeString(input));
  return extractStringField(readString(offset), "text");
}

export function close(socketId: string): void {
  const input = "{\"socketId\":" + quote(socketId) + "}";
  peekoo_websocket_close(writeString(input));
}

File: crates/peekoo-plugin-sdk/src/websocket.rs
//! WebSocket helpers for plugins.
//!
//! Requires the `net:websocket` permission.
//! Hosts must also be allowlisted in `peekoo-plugin.toml`.

use extism_pdk::{Error, Json};

use crate::host_fns::{
    peekoo_websocket_close, peekoo_websocket_connect, peekoo_websocket_recv, peekoo_websocket_send,
    WebSocketCloseRequest, WebSocketConnectRequest, WebSocketRecvRequest, WebSocketSendRequest,
};

/// Opens a new WebSocket connection and returns a host-managed socket id.
pub fn connect(url: &str) -> Result<String, Error> {
    let response = unsafe {
        peekoo_websocket_connect(Json(WebSocketConnectRequest {
            url: url.to_string(),
        }))?
    };
    Ok(response.0.socket_id)
}

/// Sends a text frame on an existing socket.
pub fn send(socket_id: &str, text: &str) -> Result<(), Error> {
    unsafe {
        peekoo_websocket_send(Json(WebSocketSendRequest {
            socket_id: socket_id.to_string(),
            text: text.to_string(),
        }))?
    };
    Ok(())
}

/// Receives the next text message for an existing socket.
pub fn recv(socket_id: &str) -> Result<String, Error> {
    let response = unsafe {
        peekoo_websocket_recv(Json(WebSocketRecvRequest {
            socket_id: socket_id.to_string(),
        }))?
    };
    Ok(response.0.text)
}

/// Closes an existing socket and releases the host-managed handle.
pub fn close(socket_id: &str) -> Result<(), Error> {
    unsafe {
        peekoo_websocket_close(Json(WebSocketCloseRequest {
            socket_id: socket_id.to_string(),
        }))?
    };
    Ok(())
}
